kotlin-codegen/compiler/src/main/java/com/squareup/moshi/TypeResolver.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.ParameterizedTypeName
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.kotlinpoet.TypeVariableName
+import com.squareup.kotlinpoet.WildcardTypeName
+
+/**
+ * Resolves type parameters against a type declaration. Use this to fill in type variables with
+ * their actual type parameters.
+ */
+open class TypeResolver {
+  open fun resolveTypeVariable(typeVariable: TypeVariableName): TypeName = typeVariable
+
+  fun resolve(typeName: TypeName): TypeName {
+    return when (typeName) {
+      is ClassName -> typeName
+
+      is ParameterizedTypeName -> {
+        ParameterizedTypeName.get(
+            typeName.rawType, *(typeName.typeArguments.map { resolve(it) }.toTypedArray()))
+      }
+
+      is WildcardTypeName -> {
```

Comment:
existing behavior, unrelated to the PR:
i wonder if these types are worth supporting. i'm unsure of the use case.

Reply:
They come up invisibly in Kotlin’s immutable List type.